Support locking windows

johnwparent opened the pull request:
Staging area for Windows locking support spun off from #27021

johnwparent committed:
Windows pywin32 locking

Locking ranges is not currently supported on Windows.

pywintypes.OVERLAPPED in the Windows lock implementation may
not be a class member (#22244)

Windows: Lock entire file even if lock-by-range is requested (#24183)

Windows lock timeout (#25189)

Fixup improper cross-platform lock catching

Add locking test to Windows CI (#25233)

Co-authored-by: lou.lawrence@kitware.com <lou.lawrence@kitware.com>

johnwparent committed:
Add locking Windows actions

johnwparent committed:
Update locks for win

scheibelp commented on .github/workflows/windows_python.yml:
```diff
@@ -111,10 +111,30 @@ jobs:
+    - name: Build Test
+      run: |
+        spack compiler find
+        echo F|xcopy .\spack\share\spack\qa\configuration\windows_locking_config.yaml $env:USERPROFILE\.spack\windows\config.yaml
```
IMO as part of this PR we can just make the `buildtest` job do locking (i.e. we don't need a separate `buildtest` and `locking-buildtest` job).

scheibelp commented on lib/spack/llnl/util/lock.py:
```diff
@@ -6,18 +6,23 @@
+    import win32con
+    import win32file
 
+    import pywintypes  # isort:skip
```
(minor) would the `isort:skip` be needed if this were placed just above `import win32con`?

scheibelp commented on lib/spack/spack/llnl/util/lock.py:
```diff
@@ -232,6 +270,7 @@ def __init__(self, path, start=0, length=0, default_timeout=None,
         """
         self.path = path
         self._file = None
+        self._file_mode = ""
```
This variable is never read and only appears to ever be set to `""`.

johnwparent replied:
Vestigial code from 2021 😬

scheibelp commented on lib/spack/spack/llnl/util/lock.py:
```diff
@@ -346,51 +400,63 @@ def _poll_lock(self, op):
         """Read PID and host data out of the file if it is there."""
+        if is_windows:
+            # Not implemented for windows
+            return
```
If this function (and `_write_log_debug_data`) are not run at all on Windows, then IMO it's better to not call them at all in `_poll_lock`.

scheibelp commented on lib/spack/spack/llnl/util/lock.py:
```diff
@@ -169,29 +174,62 @@ def _attempts_str(wait_time, nattempts):
+            try:
+                suc = func(self, *args, **kwargs)
+            except Exception as e:
+                if self._current_lock:
```
Note that `LockType.READ` is `0` on Windows, which would evaluate to `False` (the same as `None`) - is that intended?

Also, is `_current_lock` possibly redundant with checking `self._writes` and `self._reads`? I think it might be worthwhile to create a property which examines those two attributes (which would have the same effect as `_current_lock` but would avoid duplicating the bookkeeping).

scheibelp commented on lib/spack/spack/llnl/util/lock.py:
```diff
@@ -254,6 +293,19 @@ def __init__(self, path, start=0, length=0, default_timeout=None,
+        # Mapping of supported locks to description
+        self._current_lock = None
+
+    def __lock_fail_condition(self, e):
```
(minor: style) we generally don't prefix with `__` (just a single `_` for private methods).

scheibelp commented on lib/spack/spack/llnl/util/lock.py:
```diff
@@ -282,7 +334,8 @@ def __repr__(self):
         rep = '{0}('.format(self.__class__.__name__)
         for attr, value in self.__dict__.items():
-            rep += '{0}={1}, '.format(attr, value.__repr__())
+            if attr != "LOCK_CATCH":
```
It looks like this is no longer an attribute on this object (all those moved to `LockType`).

scheibelp commented on lib/spack/spack/llnl/util/lock.py:
```diff
@@ -346,51 +400,63 @@ def _poll_lock(self, op):
-            # EAGAIN and EACCES == locked by another process (so try again)
-            if e.errno not in (errno.EAGAIN, errno.EACCES):
+        except LockType.LOCK_CATCH as e:
+            # check if lock failure or lock is already held
```
(minor: wording) this says what we're looking for but not why we're doing it (i.e. is it good if this is the case, or bad?). Suggestion:

> Ignore exceptions related to the lock already being held (since that is what this object is managing).

scheibelp commented on lib/spack/spack/llnl/util/lock.py:
```diff
@@ -512,7 +594,7 @@ def downgrade_write_to_read(self, timeout=None):
         timeout = timeout or self.default_timeout
 
-        if self._writes == 1 and self._reads == 0:
+        if self._writes == 1:
```
It seems good to remove this check, but if we are multiply-nested inside of read locks, I'm worried about a few lines below this where we set `self._reads= 1`. My first thought is that we should actually do `self._reads += 1` (which I think is correct because `acquire_write()` doesn't increment `self._reads`.

scheibelp commented on lib/spack/spack/llnl/util/lock.py:
```diff
@@ -531,7 +613,7 @@ def upgrade_read_to_write(self, timeout=None):
         timeout = timeout or self.default_timeout
 
-        if self._reads == 1 and self._writes == 0:
+        if self._reads >= 1 and self._writes == 0:
```
(analogous concern to `downgrade_write_to_read`)

scheibelp reviewed (changes requested):
I have some preliminary comments/requests.

In addition, it looks like the style checker is currently complaining.

alecbcs commented:
@johnwparent do you have plans to come back to this PR or should we close?

johnwparent committed:
Remove locking guard on Windows

johnwparent commented:
> > @johnwparent do you have plans to come back to this PR or should we close?
> 
> I have plans to come back to this concept pending a conversation about expected behavior on Windows vs Linux. That said, IIRC the lock module has been refactored pretty heavily and these changes are very out date. I think we're ok to close this and replace with an issue to track locking on Windows.
> 
> Thanks for churning through the MR backlog @alecbcs!

@alecbcs Good thing we didn't close this, I actually need it now 😁

johnwparent committed:
Style

johnwparent committed:
Fixup import

johnwparent committed:
wip

johnwparent committed:
implement partial locking

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Switchover locking

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Remove old lock upgrade approach

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
win32file locking test

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Support ranges properly

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
style

Signed-off-by: John Parent <john.parent@kitware.com>

scheibelp commented on lib/spack/spack/llnl/util/lock.py:
```diff
@@ -267,6 +384,18 @@ def __init__(
+    def __lock_fail_condition(self, e):
+        if IS_WINDOWS:
+            # 33 "The process cannot access the file because another
+            #     process has locked a portion of the file."
```
(question) is this true for any portion, or only if an overlapping portion is locked?

johnwparent replied:
Only when the section lock prevents concurrent access.

scheibelp commented on lib/spack/spack/llnl/util/lock.py:
```diff
@@ -392,12 +532,15 @@ def _poll_lock(self, op: int) -> bool:
             # EAGAIN and EACCES == locked by another process (so try again)
             if e.errno not in (errno.EAGAIN, errno.EACCES):
                 raise
+        except LockType.LOCK_CATCH as e:
```
It looks like `LOCK_CATCH/lock_fail_condition` were designed to handle both of these `except` blocks - is the prior one duplicated?

scheibelp commented on lib/spack/spack/llnl/util/lock.py:
```diff
@@ -172,9 +191,106 @@ def _attempts_str(wait_time, nattempts):
+    range, the exclusive lock must also be unlocked and must be unlocked *first*
+    From the MSFT docs:
+
+    A shared lock can overlap an exclusive lock if both locks were created using
```
(question) what does it mean for a region to have a shared lock and exclusive lock? I assume they must both be held by the same process?

johnwparent replied:
You can have a shared (read) and exclusive (write) lock on the same file region at the same time, but they must be held not only by the same process, but by the same handle.

scheibelp commented on lib/spack/spack/llnl/util/lock.py:
```diff
@@ -172,9 +191,106 @@ def _attempts_str(wait_time, nattempts):
+    so we just give it up
+    """
+    timeout = timeout or lock.default_timeout
+    # release the read lock first
```
(question) I don't see that happening here

scheibelp commented on lib/spack/spack/llnl/util/lock.py:
```diff
@@ -172,9 +191,106 @@ def _attempts_str(wait_time, nattempts):
+        LOCK_EX = fcntl.LOCK_EX
+        LOCK_SH = fcntl.LOCK_SH
+        LOCK_NB = fcntl.LOCK_NB
+        LOCK_UN = fcntl.LOCK_UN
```
Not sure if `LOCK_UN` should be defined in this type if it isn't exposed on all platforms (although not strictly an error: I only see `LockType.LOCK_UN` accessed on unix).

johnwparent replied:
You are correct, its unix only.
Could just as easily be left as `fcntl.LOCK_UN` or the whole interface refactored into more atomic components. I don't really have a strong opinion.

scheibelp commented on lib/spack/spack/test/llnl/util/lock.py:
```diff
@@ -583,7 +585,7 @@ def test_write_lock_timeout_with_multiple_readers_3_2_ranges(lock_path):
 
 
-@pytest.mark.skipif(getuid() == 0, reason="user is root")
+@pytest.mark.skipif(sys.platform == "win32" or getuid() == 0, reason="user is root")
```
I think the `reason` needs to be updated here (I'm assuming it doesn't apply to the `sys.platform` check).

johnwparent replied:
Yup! (it does not)

scheibelp commented on lib/spack/spack/test/llnl/util/lock.py:
```diff
@@ -1347,10 +1351,37 @@ def _lockf(fd, cmd, len, start, whence):
+
+        monkeypatch.setattr(win32file, "LockFileEx", LockFileEx)
+
+        if err_num in [errno.EAGAIN, errno.EACCES]:
```
I would have expected that the relevant checks on Windows are `32/33`, not `errno.EAGAIN/EACCES` - are both these paths taken for the tests on Windows?

johnwparent replied:
I'd say they are definitely not. Will update (I admittedly don't recall updating this test after my last round of updates)

johnwparent replied:
This test case is skipped on Windows, there's an analog beneath it.

scheibelp commented on lib/spack/spack/test/llnl/util/lock.py:
```diff
@@ -327,7 +329,7 @@ def __call__(self, barrier):
 
 
 class TimeoutWrite:
-    def __init__(self, lock_path, start=0, length=0):
```
(question) Is a timeout length of 0 invalid on Windows? Can it be made OS-dependent?

johnwparent replied:
I believe 0 is valid, yes, but will make OS dependent.

scheibelp commented on lib/spack/spack/llnl/util/lock.py:
```diff
@@ -172,9 +191,106 @@ def _attempts_str(wait_time, nattempts):
+    This gate file prevents any other process/lock from catching
+    the lock with a competing write during the release part of the upgrade.
+    """
+    # probably something like foo.lock.lock
```
Not sure what this comment means. If it means "the lock used for exclusion locking is based on the lock name and adds `.lock`, which usually means it will look like `...lock.lock`"; if that's what it means, I think expanding the comment would be useful. I think clarifying the lock purpose with a distinct suffix would also be useful (so if people see this file lying around after an error, it's easier to find it in the source).

scheibelp commented on lib/spack/spack/llnl/util/lock.py:
```diff
@@ -172,9 +191,106 @@ def _attempts_str(wait_time, nattempts):
+    try:
+        # don't use the acquire lock method to avoid
+        # recursion
+        lk._lock(LockType.WRITE, timeout=timeout)
```
(need to think about this more but) I think for this to work, write locks *also* need to acquire this lock (to be clear, I mean that entities that skip read and go straight to write need to acquire this gate lock as well), otherwise between lines 238 and 239, another process can acquire a non-upgrade write lock, write to the data protected by this lock, and we don't have the upgrade semantics we want (I think the general use of the upgrade lock is the assumption that when you have acquired the write lock, that everything you read when you had the read lock is still authoritative).

johnwparent replied:
Looks like I was doing that in practice but the docstring is inaccurate for the reasons you point out. Updated.

scheibelp commented on lib/spack/spack/llnl/util/lock.py:
```diff
@@ -172,9 +191,106 @@ def _attempts_str(wait_time, nattempts):
+    # operation unlocks the shared lock.
+    #
+    # This will remove the exclusive lock
+    lock._release_lock()
```
I think for this to be safe, it needs to only release the exclusive lock if we successfully acquired the read lock, I think putting this action in a context manager makes it happen no matter what. I think the easiest way to resolve this is to put the "release-exclusive" operation right after the "acquire-read" operation

scheibelp reviewed (commented):
Mostly questions, I have a couple concerns

I'll note that there appears to be only one "customer" of `upgrade_read_to_write`, in `installer.py`, which may be dropped for `new_installer.py` (whatever PR does that may remove this upgrade logic). So I think it's worth figuring out the timeline/intent of that logic before focusing on the Windows-analog upgrade path here.

johnwparent committed:
cleanup

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
fixup locking class for windows upgrades

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
fixup tests

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Ensure timeouts are respected

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Start should be 0

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
wip

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
overhaul locking design

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
rm old lockfile after rebase

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
restore conftest comment for stage files to be ignored

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Migrate core locking logic and test to unix

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
reorganize lock testing

Shared backend logic goes into relevant testing module, leaving lock_unix and lock_windows to test platform specific locking behavior

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Final refinement

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
style

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
refactor module

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
cleanup imports

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
I think that should make mypy happy?

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
large refactor, just to make mypy not gripe

Signed-off-by: John Parent <john.parent@kitware.com>
